lib/base64.h

```diff
@@ -118,6 +159,41 @@ namespace macaron {
             }
         }
 
+    public:
+
+        static std::string Encode(const std::vector<uint8_t>& data) {
+            return Encode(data, base64EncodingTable, false);
+        }
+
+        static std::string EncodeBase64URL(const std::vector<uint8_t>& data) {
+            return Encode(data, base64URLEncodingTable, true);
+        }
+
+        static void Decode(const std::string& input, std::vector<uint8_t>& out) {
+            if (input.size() % 4 != 0)
+            {
+                throw "invalid base64 string length";
+            }
+            return Decode(input, base64DecodingTable, out);
+        }
+
+        static void DecodeBase64URL(const std::string& input, std::vector<uint8_t>& out) {
+            int inputSizeMod4 = input.size() % 4;
+            //it's within spec https://www.rfc-editor.org/rfc/rfc4648#section-5
+            if(inputSizeMod4 != 0)
+            {
+                std::string input2(input);
```

Comment:
given we're copying the string anyway, can consider doing a string replace with _ and -, and then you don't need the base64URL tables

Reply:
i have adding more unit tests so i think won't need to do this inefficient string replace

Reply:
right, just to note that this string copy usually requires a memory allocation, which can be much slower than a string replace